core/domain/search_services.py

```diff
@@ -0,0 +1,201 @@
+    return rights.status != rights_manager.ACTIVITY_STATUS_PRIVATE
+
+
+def _should_index_collection(collection_summary):
+    """Returns whether the collection corresponding to the given
+    collection summary should be indexed for future search queries.
+
+    Args:
+        collection_summary: Collection summary object.
+    """
+    rights = rights_manager.get_collection_rights(collection_summary.id)
+    return rights.status != rights_manager.ACTIVITY_STATUS_PRIVATE
+
+
+def _exp_summary_to_search_dict(exp_summary):
+    """Updates the dict to be returned, whether the given exploration is to
+    be indexed for further queries or not.
+
+    Args:
+        exp_summary: ExplorationSummary. ExplorationSummary domain object.
+
+    Returns:
+        dict. The representation of the given exploration, in a form that can
+        be used by the search index.
+    """
+    rights = rights_manager.get_exploration_rights(exp_summary.id)
+    doc = {'id': exp_summary.id,
+           'language_code': exp_summary.language_code,
+           'title': exp_summary.title,
+           'category': exp_summary.category,
+           'tags': exp_summary.tags,
+           'objective': exp_summary.objective,
+           'rank': get_search_rank_from_exp_summary(exp_summary)
+          }
+    doc.update(_exp_rights_to_search_dict(rights))
+    return doc
+
+
+def _collection_summary_to_search_dict(collection_summary):
+    """Converts a collection domain object to a search dict.
+
+    Args:
+        collection_summary: Collection. The CollectionSummary domain object to
+        be converted.
+
+    Returns:
+        dict. The representation of the given collection, in a form that can
+        be used by the search index.
+    """
+    rights = rights_manager.get_collection_rights(collection_summary.id)
+    doc = {'id': collection_summary.id,
+           'title': collection_summary.title,
+           'category': collection_summary.category,
+           'objective': collection_summary.objective,
+           'language_code': collection_summary.language_code,
+           'tags': collection_summary.tags,
+           'rank': get_collection_search_rank(collection_summary.id)
+          }
+    doc.update(_collection_rights_to_search_dict(rights))
+    return doc
+
+
+def get_collection_search_rank(collection_id):
+    """Gets the search rank of a given collection.
+
+    Args:
+        collection_id: str. ID of the collection whose rank is to be retrieved.
+
+    Returns:
+        int. An integer determining the document's rank in search.
+
+        Featured collections get a ranking bump, and so do collections that
+        have been more recently updated.
+    """
+    rights = rights_manager.get_collection_rights(collection_id)
+    rank = _DEFAULT_RANK + (_STATUS_PUBLICIZED_BONUS if
+                            rights.status == rights_manager
+                            .ACTIVITY_STATUS_PUBLIC else 0)
```

Comment:
OK, something's really wrong here ... ACTIVITY_STATUS_PUBLIC means something completely different from ACTIVITY_STATUS_PUBLICIZED, so this whole clause has been removed in develop, I think.

I'm starting to get a bit concerned that, because this PR has been around for a while, it is operating based on assumptions that were present in an older version of the codebase -- and I'm worried that trying to fix those assumptions on a case-by-case basis is like playing wack-a-mole, and we'll see odd regressions in production if this gets merged. So, I'd like to suggest that, now it's clear how to do the refactor -- would it be possible to close this PR, sync to the latest version of develop, and start a new one following exactly the same approach, but using the current version of the code instead (i.e. don't copy and paste from this PR unless it's new code that's been added specifically for this PR, and even then, double check that it still makes sense with the current codebase)? That way, we will preserve the current invariants and not break anything.